Allow overriding default event namespace in Kubelet.

hoskeri opened the pull request:
#### What this PR does / why we need it:

Kubelet uses the 'default' namespace to post events about itself.

Cluster operators may want to implement restrictions on the 'default' namespace or discourage its use. This would result in errors in Kubelet, and the loss of those events.

Allow overriding the namespace Kubelet posts events to. The default behavior is unchanged.

/kind feature

```release-note
kubelet: allow overriding default event namespace with --event-namespace.
```

NOTE: Will add tests once I get some feedback from reviewers that this makes sense to do.

mtaufen commented:
/ok-to-test

lmktfy commented:
> `kubelet: allow overriding default event namespace with --event-namespace`

Configuring the kubelet using command line arguments is deprecated. Can we swap that to a configuration parameter.

tallclair commented:
Ha, this TODO from pre-1.0 lives on:
https://github.com/kubernetes/kubernetes/blob/bbd859808d0cef566fbfc89639926b7cc50d9bd5/staging/src/k8s.io/kubectl/pkg/describe/describe.go#L3697
Originally from
https://github.com/kubernetes/kubernetes/pull/6026/files#diff-7b1d644fa9c0f272ece4e97e45bc2098daace4ef187c075dcbed6dc367ef4351R374

Anyhow, looks like this would break listing node events through describe.

> defining that kube-node-events is a system namespace, and

Why a new namespace? Why not just reuse `kube-system`? 

Is there anywhere else that we record events for cluster-scoped objects? If so, we should probably maintain consistency with those.

aaron-prindle commented:
/remove-sig api-machinery

aaron-prindle commented:
/remove-sig api-machinery

hoskeri committed:
Support overriding namespace in event recorder

Events involving resources without a namespace are published to the the
'default' namespace.  Add an option to send such events to a namespace
other than default.

hoskeri committed:
Drop unused references to eventRecorder

hoskeri committed:
Allow overriding default event namespace in Kubelet.

Kubelet uses the "default" namespace to post events
about itself.

Cluster operators may want to implement restrictions on the "default"
namespace or discourage its use. This would result in errors in Kubelet,
and the loss of those events.

Allow overriding the namespace Kubelet posts events to.
The default behavior is unchanged.

Requires the KubeletEventsNamespace feature gate.

dims commented:
> @dims I don't think so. Are you asking for a KEP?

@mtaufen i'll leave that to sig-node chairs/leads, i was curious and could not find one, hence asking if there is/was one.

hoskeri commented:
/remove-lifecycle stale

I still want to pursue this.

aaron-prindle commented:
/remove-sig api-machinery

tallclair commented:
/assign
